src/storage/src/source/reclock/compat.rs

```diff
@@ -311,111 +324,124 @@ impl RemapHandle for PersistHandle {
-                            new_native_frontier
-                                .into_iter()
-                                .map(|src_ts| (src_ts, ts, 1)),
-                        );
+                    if self.partitioned_source {
+                        native_frontier_updates =
+                            self.synthesize_empty_ranges(native_frontier_updates);
                     }
-                    // Then, consolidate the native updates and we're done
-                    consolidation::consolidate_updates(&mut native_frontier_updates);
-
-                    // Finally, apply the updates to our local view of the native frontier
-                    self.native_source_upper.update_iter(
-                        native_frontier_updates
-                            .iter()
-                            .map(|(src_ts, _ts, diff)| (src_ts.clone(), *diff)),
-                    );
-                    self.upper = new_upper.clone();
 
                     return Some((native_frontier_updates, new_upper));
                 }
                 ListenEvent::Updates(msgs) => {
                     for ((update, _), ts, diff) in msgs {
-                        let binding = unpack_binding(update.expect("invalid row"));
+                        let binding = Self::FromTime::try_from(update.expect("invalid row"))
+                            .expect("invalid binding");
                         self.pending_batch.push((binding, ts, diff));
                     }
                 }
             }
         }
         None
     }
+}
+
+#[async_trait::async_trait(?Send)]
+impl RemapHandle for PersistHandle {
+    async fn compare_and_append(
+        &mut self,
+        updates: Vec<(Self::FromTime, Self::IntoTime, Diff)>,
+        upper: Antichain<Self::IntoTime>,
+        new_upper: Antichain<Self::IntoTime>,
+    ) -> Result<(), Upper<Self::IntoTime>> {
+        let row_updates = updates.into_iter().filter_map(|(partitioned, ts, diff)| {
+            match partitioned.interval() {
+                // If we are not dealing with a partitioned source and the
+                // confusingly-named partition() function is none, it means the
+                // underlying partition is an interval, which is serialized a
+                // range. We do not encode ranges in non-partitioned sources, so
+                // drop it on the floor.
+                //
+                // Not that it is because of this that `Self::next` has to act
+                // as if it's read intervals, though it hasn't.
+                Interval::Range(_, _) if !self.partitioned_source => None,
+                // These ranges would be converted to empty in the conversion to
+                // `SourceData`, so filter them out.
+                Interval::Range(
+                    RangeBound::Elem(PartitionId::Kafka(l)),
+                    RangeBound::Elem(PartitionId::Kafka(u)),
+                ) if u - l == 1 => None,
```

Comment:
Whether empty ranges are included should be a decision to whoever is generating the frontiers so we shouldn't filter them out here. Currently the conversions are dumb and generate them so we should also write them down. When sources become smarter and generate saner partitions they will self-correct as the frontier will progress

Reply:
Sounds good--I adjusted the code to write them into the shards and filter them out as necessary.